pyccel/parser/scope.py

```diff
@@ -0,0 +1,325 @@
+        """
+        Create a new child Scope object which has the current object as parent.
+
+        The parent scope can access the child scope through the '_sons_scopes'
+        dictionary, using the provided name as key. Conversely, the child scope
+        can access the parent scope through the 'parent_scope' attribute.
+
+        Parameters
+        ----------
+        name : str
+            Name of the new scope, used as a key to retrieve the new scope.
+
+        kwargs : dict
+            Keyword arguments passed to __init__() for object initialization.
+
+        Returns
+        -------
+        child : Scope
+            New child scope, which has the current object as parent.
+
+        """
+
+        child = Scope(**kwargs, parent_scope = self)
+
+        self._sons_scopes[name] = child
+        child.parent_scope = self
+
+        return child
+
+    @property
+    def imports(self):
+        """ A dictionary of objects imported in this scope
+        """
+        return self._imports
+
+    @property
+    def variables(self):
+        """ A dictionary of variables defined in this scope
+        """
+        return self._locals['variables']
+
+    @property
+    def classes(self):
+        """ A dictionary of classes defined in this scope
+        """
+        return self._locals['classes']
+
+    @property
+    def functions(self):
+        """ A dictionary of functions defined in this scope
+        """
+        return self._locals['functions']
+
+    @property
+    def macros(self):
+        """ A dictionary of macros defined in this scope
+        """
+        return self._locals['macros']
+
+    @property
+    def headers(self):
+        """A dictionary of user defined headers which may
+        be applied to functions in this scope"""
+        return self._locals['headers']
+
+    @property
+    def templates(self):
+        """A dictionary of user defined templates which may
+        be applied to functions in this scope"""
+        return self._locals['templates']
+
+    @property
+    def decorators(self):
+        """Dictionary of Pyccel decorators which may be
+        applied to a function definition in this scope."""
+        return self._locals['decorators']
+
+    @property
+    def static_functions(self):
+        """ A dictionary of static functions defined in this scope
+        """
+        return self._locals['static_functions']
+
+    @property
+    def cls_constructs(self):
+        """ A dictionary of datatypes for the classes defined in
+        this scope
+        """
+        return self._locals['cls_constructs']
+
+    @property
+    def sons_scopes(self):
+        """ A dictionary of all the scopes contained within the
+        current scope
+        """
+        return self._sons_scopes
+
+    @property
+    def symbolic_functions(self):
+        """ A dictionary of symbolic functions defined in this scope
+        """
+        return self._locals['symbolic_functions']
```

Comment:
@saidctb Do you know what a symbolic function is?

Reply:
Is this related to issue #330 ?

Reply:
I see this seems to be for the `@sympy` and `@python` decorators as well as lambdas

Reply:
yes this is for symbolic functions, `@sympy` and lambdas, `@python` should have a different property

Reply:
Ah, ok. It looks like I removed the `@python` version : a4766d2
It wasn't used anywhere in the code. Probably this falls under issue #668

Reply:
@yguclu @saidctb we need to decide what we want to do with issue #668 in order to finish this PR. If we want to remove `@python` then it should be done completely. If we want to keep it then I need to put the `python_functions` category back

Reply:
I think we can remove it completely for now. We will reconsider it in the future.

Reply:
If I remove this then I should also clean up the places where it is used which means for example removing all code related to Lambdas. Lambdas don't work currently so this may not be a problem. I just want to make sure before removing it

Reply:
Fixed:
[5be3529](https://github.com/pyccel/pyccel/pull/1068/commits/5be3529e0e105b0ac7db11c13ac791354c1727e5)
[47ae7e1](https://github.com/pyccel/pyccel/pull/1068/commits/47ae7e14c7a982ca972dd335019f1a3974808199)